File: src/main/java/twopointers/ThreeSum.py
from typing import List


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        result = []
        sorted_nums = sorted(nums)
        # since x + y + z = 0 -> x + y = -z -> twoSum with target = -z
        for z_index in range(len(sorted_nums) - 2):
            # skip duplicate z
            if z_index > 0 and sorted_nums[z_index] == sorted_nums[z_index - 1]:
                continue
            self.two_sum(z_index + 1, len(nums) - 1, sorted_nums, -sorted_nums[z_index], result)
        return result

    def two_sum(self, x_index: int, y_index: int, nums: List[int], target: int, result: List[List[int]]):
        while x_index < y_index:
            if nums[x_index] + nums[y_index] == target:
                result.append([-target, nums[x_index], nums[y_index]])
                x_index += 1
                while x_index < y_index and nums[x_index] == nums[x_index - 1]:
                    x_index += 1
                # skip duplicate
                y_index -= 1
                while x_index < y_index and nums[y_index] == nums[y_index + 1]:
                    y_index -= 1
            elif nums[x_index] + nums[y_index] < target:
                x_index += 1
            elif nums[x_index] + nums[y_index] > target:
                y_index -= 1






if __name__ == '__main__':
    def test():
        return 1, 2

    result = []
    result.append([0, test()])
    print(result)

File: src/main/java/ThreeSum.java
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

class ThreeSum {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> result = new LinkedList<>();
        Arrays.sort(nums);
        // assume x + y + z = 0 -> y + z = -x
        for (int xIndex = 0; xIndex < nums.length - 2; xIndex++) {
            final int x = nums[xIndex];
            // to skip same x
            if (hasSameX(nums, xIndex)) {
                continue;
            }
            twoSum(nums, -x, xIndex + 1, result);
        }
        return result;
    }

    public boolean hasSameX(int[] nums, int xIndex) {
        return xIndex >= 1 && nums[xIndex] == nums[xIndex-1];
    }

    public void twoSum(int[] nums, int target, int headIndex, List<List<Integer>> result) {
        int yIndex = headIndex;
        int zIndex = nums.length - 1;
        while (yIndex < zIndex) {
            final int currentSum = nums[yIndex] + nums[zIndex];
            if (currentSum == target) {
                int x = -target;
                int y = nums[yIndex];
                int z = nums[zIndex];
                result.add(List.of(x, y, z));
                yIndex++;
                zIndex--;
                // skip duplicate y
                while (yIndex < zIndex && nums[yIndex] == nums[yIndex - 1]) {
                    yIndex++;
                }
                // skip duplicate z
                while (yIndex < zIndex && nums[zIndex] == nums[zIndex + 1]) {
                    zIndex--;
                }

            } else if (currentSum < target) {
                yIndex++;
            } else if (currentSum > target) {
                zIndex--;
            }
        }
    }
}
